File: src/events.js
'use strict';

class EventEmitter {
  constructor() {
    this._listeners = new Map();
    this._once = new Set();
  }

  on(event, fn) {
    if (!this._listeners.has(event)) {
      this._listeners.set(event, []);
    }
    this._listeners.get(event).push(fn);
    return this;
  }

  once(event, fn) {
    this._once.add(fn);
    return this.on(event, fn);
  }

  off(event, fn) {
    if (!this._listeners.has(event)) return this;
    if (fn) {
      const fns = this._listeners.get(event).filter((f) => f !== fn);
      this._listeners.set(event, fns);
    } else {
      this._listeners.delete(event);
    }
    return this;
  }

  emit(event, ...args) {
    const fns = this._listeners.get(event) || [];
    const wildcard = this._listeners.get('*') || [];
    for (const fn of [...fns]) {
      fn(...args);
      if (this._once.has(fn)) {
        this.off(event, fn);
        this._once.delete(fn);
      }
    }
    for (const fn of [...wildcard]) {
      fn(event, ...args);
      if (this._once.has(fn)) {
        this.off('*', fn);
        this._once.delete(fn);
      }
    }
    return this;
  }

  listenerCount(event) {
    return (this._listeners.get(event) || []).length;
  }
}

module.exports = { EventEmitter };

File: python/phyloid_engine/events.py
from __future__ import annotations

from typing import Any, Callable


class EventEmitter:
    """Pub/sub event bus with wildcard support and chainable API."""

    def __init__(self) -> None:
        self._listeners: dict[str, list[Callable[..., Any]]] = {}
        self._once: set[Callable[..., Any]] = set()

    def on(self, event: str, fn: Callable[..., Any]) -> EventEmitter:
        self._listeners.setdefault(event, []).append(fn)
        return self

    def once(self, event: str, fn: Callable[..., Any]) -> EventEmitter:
        self._once.add(fn)
        return self.on(event, fn)

    def off(self, event: str, fn: Callable[..., Any] | None = None) -> EventEmitter:
        if event not in self._listeners:
            return self
        if fn:
            self._listeners[event] = [f for f in self._listeners[event] if f is not fn]
        else:
            del self._listeners[event]
        return self

    def emit(self, event: str, *args: Any) -> EventEmitter:
        fns = list(self._listeners.get(event, []))
        wildcard = list(self._listeners.get("*", []))
        for fn in fns:
            fn(*args)
            if fn in self._once:
                self.off(event, fn)
                self._once.discard(fn)
        for fn in wildcard:
            fn(event, *args)
            if fn in self._once:
                self.off("*", fn)
                self._once.discard(fn)
        return self

    def listener_count(self, event: str) -> int:
        return len(self._listeners.get(event, []))
